components/constellation/constellation.rs

```diff
@@ -1945,6 +1945,12 @@ where
                     EmbedderMsg::MediaSessionEvent(event),
                 ));
             },
+            FromScriptMsg::StartWebGPU(sender) => {
+                self.webgpu = WebGPU::new();
```

Comment:
Even though you are tracking the `webgpu_channel_requested` flag inside the DOM GPU, multiple script-threads(`EventLoop` from the perspective of the constellation) could send this message, in which case you would be overwriting previously set `self.webgpu`(and running multiple gpu threads?). 

Also, even within one script-thread, multiple windows could be running on it(see `documents` in script-thread), in which case I think you could also have multiple DOM `GPU` instances, each tracking their own `webgpu_channel_requested` flag. 

So I think you need to decide:

1. Do you want to allow multiple DOM `GPU` to share a `WebGPU` at the same time? Is so, how would they be shared(one webgpu shared across the browser, one shared per origin, per agent-cluster?). Then the `StartWebGPU` would need to include information so that you can either clone an existing one, or start a new one.
2. Do you want to allow multiple parallel `WebGPU` at all? In that case, you would need a map with a key representing how an instance is shared, so that you could get it out here, clone the sender, or start a new one. 

It's not clear to me from the spec how sharing should happen, although @kvark has previously said multiple parallel applications are wanted(although it's not clear if that means sharing a single webgpu, or running multiple ones, or a combination of that).

Reply:
Yes, the choice of boundaries for separating WebGPU instances is still ours.

Here is how it works in Gecko, for the reference:
  - each content process manages multiple tabs
  - a single WebGPU remote client is established per content process
  - it spawns a separate WebGPU server on the GPU process
  - the IPC is a child of CompositorBridge and thus the server message loop is happening on a thread that this parent protocol manages (I assume it's a separate thread)

Does Servo have multiple content processes? I think we can start here by having the same approach - one WebGPU client-server pair per content process.

Reply:
Yes so a content-process would correspond to a [`EventLoop`](https://github.com/servo/servo/blob/master/components/constellation/event_loop.rs) in Servo. 

From the constellation's perspective a document corresponds to a [`Pipeline`](https://github.com/servo/servo/blob/master/components/constellation/pipeline.rs), of which multiple can run on one event-loop. 

When a pipeline is spawned, an event-loop is either created or re-used, see https://github.com/servo/servo/blob/6e3c131139c412031a4d288aa1ffd048ca0311c3/components/constellation/pipeline.rs#L219

The "re-using" of event-loops is done when a pipeline is part of the same [`BrowsingContextGroup`](https://github.com/servo/servo/blob/6e3c131139c412031a4d288aa1ffd048ca0311c3/components/constellation/constellation.rs#L233), and based on the `Host` of the url of that pipeline, see [`get_event_loop`](https://github.com/servo/servo/blob/6e3c131139c412031a4d288aa1ffd048ca0311c3/components/constellation/constellation.rs#L1024).

So, this would actually be the first time something is shared like that on an "per-event-loop" basis, and you're going to have to provide some infra for this, because currently the constellation receives message from a given pipeline(see how each `ScriptMsg` comes with a `PipelineID`, representing the pipeline from which the message originated).

I think you could do something like:

1. As suggested in the other comment, make requesting an adapater a `ScriptMsg`.
2. When you receive it:
    - get the corresponding pipeline 
    - get the top-level browsing context for that pipeline.
    - get the host for that pipeline.
3. The above info should be enough to then grab the right webgpu instance. 
4. I would suggest you store them on a `HashMap<Host, WebGPU> inside a `BrowsingContextGroup`, similarly to what is done with `EventLoop`.